File: src/main/java/com/evergreenclient/hudmod/utils/HttpsUtils.java
/*
 * Copyright (C) Evergreen [2020 - 2021]
 * This program comes with ABSOLUTELY NO WARRANTY
 * This is free software, and you are welcome to redistribute it
 * under the certain conditions that can be found here
 * https://www.gnu.org/licenses/lgpl-3.0.en.html
 */

package com.evergreenclient.hudmod.utils;

import com.evergreenclient.hudmod.EvergreenHUD;

import java.io.*;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLConnection;

public class HttpsUtils {

    public static void downloadFile(URL url, File destination) throws Exception {
        if (!destination.exists()) {
            URLConnection con = url.openConnection();
            DataInputStream dis = new DataInputStream(con.getInputStream());
            byte[] data = new byte[con.getContentLength()];

            for (int i = 0; i < data.length; i++) {
                data[i] = dis.readByte();
            }

            dis.close();
            FileOutputStream fos = new FileOutputStream(destination);
            fos.write(data);
            fos.close();
        }
    }

    public static String getString(String url) {
        try {
            HttpURLConnection httpClient =
                    (HttpURLConnection) new URL(url).openConnection();
            httpClient.setRequestMethod("GET");
            httpClient.setUseCaches(false);
            httpClient.setRequestProperty("User-Agent", "Evergreen/" + EvergreenHUD.VERSION);
            httpClient.setReadTimeout(15000);
            httpClient.setConnectTimeout(15000);
            httpClient.setDoOutput(true);

            try (BufferedReader in = new BufferedReader(
                    new InputStreamReader(httpClient.getInputStream()))) {

                StringBuilder response = new StringBuilder();
                String line;

                while ((line = in.readLine()) != null) {
                    response.append(line);
                }

                return response.toString();

            }
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }

}

File: core/src/main/kotlin/dev/isxander/evergreenhud/utils/HttpsUtils.kt
/*
 | EvergreenHUD - A mod to improve on your heads-up-display.
 | Copyright (C) isXander [2019 - 2021]
 |
 | This program comes with ABSOLUTELY NO WARRANTY
 | This is free software, and you are welcome to redistribute it
 | under the certain conditions that can be found here
 | https://www.gnu.org/licenses/lgpl-3.0.en.html
 |
 | If you have any questions or concerns, please create
 | an issue on the github page that can be found here
 | https://github.com/isXander/EvergreenHUD
 |
 | If you have a private concern, please contact
 | isXander @ business.isxander@gmail.com
 */

package dev.isxander.evergreenhud.utils

import dev.isxander.evergreenhud.EvergreenInfo
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import java.io.File
import java.io.FileOutputStream
import java.io.IOException

fun setupRequest(url: String): Request.Builder {
    return Request.Builder()
        .url(url)
        .addHeader("User-Agent", "EvergreenHUD/${EvergreenInfo.VERSION_FULL}")
}

fun getResponse(url: String): Response? {
    try {
        val client = OkHttpClient()
        val request: Request = setupRequest(url).build()
        return client.newCall(request).execute()
    } catch (e: IOException) {
        e.printStackTrace()
    }
    return null
}

fun getRemoteBytes(url: String): ByteArray {
    return getResponse(url)!!.body!!.bytes()
}

fun getRemoteString(url: String): String {
    return getResponse(url)!!.body!!.string()
}

fun downloadFile(url: String, file: File) {
    FileOutputStream(file).use {
        it.write(getRemoteBytes(url))
    }
}